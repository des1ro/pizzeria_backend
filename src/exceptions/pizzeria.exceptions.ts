type ErrorName =
  | "PERMISION_DENIED"
  | "TABLE_ERROR"
  | "ORDER_ERROR"
  | "EMPLOYEE_ERROR";
export class PizzeriaError extends Error {
  name: string;
  message: string;
  constructor({ name, message }: { name: ErrorName; message: string }) {
    super(name);
    this.name = name;
    this.message = message;
  }
}
