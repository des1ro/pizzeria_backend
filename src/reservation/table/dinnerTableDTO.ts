export class DinnerTableDTO {
  constructor(readonly tableId: number, readonly seats: number) {}
}
