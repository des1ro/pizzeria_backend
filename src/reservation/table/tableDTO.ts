export class TableDTO {
  constructor(readonly tableId: number, readonly seats: number) {}
}
