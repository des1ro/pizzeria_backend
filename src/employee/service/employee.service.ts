import { PizzeriaError } from "../../exceptions/pizzeria.exceptions";
import { EmployeeRole } from "../enum/employee.enum";
import { EmployeeDTO } from "../model/employeeDTO";

export class EmployeeService {
  private readonly takenCheffs = new Set<EmployeeDTO>();
  constructor(
    private readonly availableCheffs = new Set<EmployeeDTO>(),
    readonly availableWaiters = new Set<EmployeeDTO>()
  ) {}
  addCheffToCrew(name: string): void {
    const cheff = new EmployeeDTO(name, EmployeeRole.Cheff);
    this.availableCheffs.add(cheff);
  }
  getCheffToOrder(): EmployeeDTO {
    const cheff = Array.from(this.availableCheffs).reverse().pop();
    if (cheff) {
      this.takenCheffs.add(cheff);
      this.availableCheffs.delete(cheff);
      return cheff;
    }
    throw new PizzeriaError({
      name: "EMPLOYEE_ERROR",
      message: "There isn't free cheffs",
    });
  }
  isCheffAvailable(): boolean {
    return this.availableCheffs.size > 0 ? true : false;
  }
  relievedCheff(cheff: EmployeeDTO) {
    if (this.takenCheffs.delete(cheff)) {
      this.availableCheffs.add(cheff);
      return;
    }
    throw new PizzeriaError({
      name: "EMPLOYEE_ERROR",
      message: "This cheff wasn't taken",
    });
  }
}
