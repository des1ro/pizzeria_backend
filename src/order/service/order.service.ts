import { EmployeeDTO } from "../../employee/model/employeeDTO";
import { PizzeriaError } from "../../exceptions/pizzeria.exceptions";
import { OrderDTO } from "../model/orderDTO";

export class OrderService {
  constructor(
    private readonly completedOrders = new Set<OrderDTO>(),
    private readonly orderQueque = new Set<OrderDTO>(),
    private readonly orderInProgress = new Map<OrderDTO, EmployeeDTO>()
  ) {}

  addOrderToQueque(order: OrderDTO): void {
    if (this.orderQueque.has(order)) {
      throw new PizzeriaError({
        name: "ORDER_ERROR",
        message: "Order is already in queque",
      });
    }
    this.orderQueque.add(order);
  }
  makeOrderInProgress(order: OrderDTO, cheff: EmployeeDTO) {
    if (this.orderQueque.delete(order)) {
      this.orderInProgress.set(order, cheff);
      return;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no such as order in queque",
    });
  }
  completeOrderAndReturnCheff(order: OrderDTO): EmployeeDTO {
    const cheff = this.orderInProgress.get(order);
    if (this.orderInProgress.has(order) && cheff) {
      this.completedOrders.delete(order);
      this.completedOrders.add(order);
      return cheff;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no order in order in progress",
    });
  }
  getOrderPrice(order: OrderDTO): number {
    return (
      (order.products.reduce((total, pizza) => total + pizza.price, 0) *
        (100 - order.discount)) /
      100
    );
  }
  getCompletedOrdersArray() {
    return Array.from(this.completedOrders);
  }
  getOrderId(): number {
    return (
      this.orderInProgress.size +
      this.orderQueque.size +
      this.completedOrders.size +
      1
    );
  }
}
