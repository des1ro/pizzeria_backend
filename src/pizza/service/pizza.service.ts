import { PizzeriaError } from "../../exceptions/pizzeria.exceptions";
import { Ingredients } from "../enum/ingredients.enum";
import { PizzaDTO } from "../model/pizzaDTO";

export class PizzaService {
  constructor(
    private readonly ingredients = new Map<Ingredients, number>(),
    private readonly pizzaList = new Set<PizzaDTO>()
  ) {}
  addIngredient(ingredient: Ingredients, amount = 1): void {
    if (this.ingredients.has(ingredient)) {
      const oldAmount = this.ingredients.get(ingredient) || 0;
      this.ingredients.set(ingredient, oldAmount + amount);
      return;
    }
    this.ingredients.set(ingredient, amount);
  }
  addPizza(pizza: PizzaDTO): void {
    this.pizzaList.add(pizza);
  }
  delatePizza(pizzaName: string): void {
    const pizza = this.findPizzaByName(pizzaName);
    this.pizzaList.delete(pizza);
  }
  getIngredientsMapOfOrder(pizzaNames: string[]) {
    const map = new Map(this.ingredients);
    for (const pizzaName of pizzaNames) {
      const pizza = this.findPizzaByName(pizzaName);
      if (pizza) {
        for (const ingredient of pizza.ingredients) {
          const value = map.get(ingredient);
          if (value === undefined || value <= 0) {
            throw new PizzeriaError({
              name: "PIZZA_SERVICE_ERROR",
              message: "Insufficient amount of ingredients",
            });
          }
          map.set(ingredient, value - 1);
        }
      }
    }
    return map;
  }
  updateIngredientsAfterOrder(map: Map<Ingredients, number>): void {
    this.ingredients.clear();
    map.forEach((value, key) => {
      this.ingredients.set(key, value);
    });
  }
  getPizzaFromNames(pizzaNames: string[]): PizzaDTO[] {
    const pizzas: PizzaDTO[] = [];
    pizzaNames.forEach((pizzaName) => {
      const pizza = this.findPizzaByName(pizzaName);
      pizzas.push(pizza);
    });
    return pizzas;
  }
  private findPizzaByName(pizzaName: string): PizzaDTO {
    const pizza = Array.from(this.pizzaList).find(
      (pizza) => pizza.name === pizzaName
    );
    if (pizza) {
      return pizza;
    }
    throw new PizzeriaError({
      name: "PIZZA_SERVICE_ERROR",
      message: "Pizza isn't on the list",
    });
  }
}
