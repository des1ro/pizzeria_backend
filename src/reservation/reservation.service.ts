import { PizzeriaError } from "../exceptions/pizzeria.exceptions";
import { TableDTO } from "../table";

export class ReservationService {
  constructor(
    private readonly freeTables = new Set<TableDTO>(),
    private readonly takenTables = new Set<TableDTO>()
  ) {}

  bookTableAndGetTableId(seats: number) {
    let tableToFind: TableDTO | undefined = undefined;
    for (const table of this.freeTables) {
      if (table.seats === seats) {
        tableToFind = table;
        break;
      }
      if (table.seats > seats) {
        if (!tableToFind || table.seats < tableToFind.seats)
          tableToFind = table;
        //Most optimal table
      }
    }
    if (tableToFind) {
      this.freeTables.delete(tableToFind);
      this.takenTables.add(tableToFind);
      return tableToFind.tableId;
    }
    throw new PizzeriaError({
      name: "TABLE_ERROR",
      message: "There is no table for so many people",
    });
  }
  setTableToAvailable(tableId: number) {
    const table = Array.from(this.takenTables).find(
      (table) => table.tableId === tableId
    );
    if (table) {
      this.takenTables.delete(table);
      this.freeTables.add(table);
      return;
    }
    throw new PizzeriaError({
      name: "TABLE_ERROR",
      message: "Wrong table id number",
    });
  }
}
