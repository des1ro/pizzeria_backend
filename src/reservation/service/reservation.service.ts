import { PizzeriaError } from "../../pizzeria/error/pizzeria.exceptions";
import { TableDTO } from "../table/tableDTO";

export class ReservationService {
  constructor(
    private readonly freeTables = new Set<TableDTO>(),
    private readonly takenTables = new Set<TableDTO>()
  ) {}
  addTable(table: TableDTO): void {
    if (!this.freeTables.has(table)) {
      this.freeTables.add(table);
      return;
    }

    throw new PizzeriaError({
      name: "RESERVATION_ERROR",
      message: "Table is already in service",
    });
  }
  getATable(seats: number) {
    let tableToFind: TableDTO | undefined = undefined;
    for (const table of this.freeTables) {
      if (table.seats === seats) {
        tableToFind = table;
        break;
      }
      if (table.seats > seats) {
        if (!tableToFind || table.seats < tableToFind.seats)
          tableToFind = table;
        //Most optimal table
      }
    }
    if (tableToFind) {
      return tableToFind;
    }
    throw new PizzeriaError({
      name: "RESERVATION_ERROR",
      message: "There is no table for so many people",
    });
  }
  bookATable(table: TableDTO): void {
    if (this.freeTables.delete(table)) this.takenTables.add(table);
  }
  setTableToAvailable(tableId: number | undefined) {
    if (tableId !== undefined) {
      const table = Array.from(this.takenTables).find(
        (table) => table.tableId === tableId
      );
      if (table) {
        this.takenTables.delete(table);
        this.freeTables.add(table);
        return;
      } else {
        throw new PizzeriaError({
          name: "RESERVATION_ERROR",
          message: "Table with the specified ID does not exist.",
        });
      }
    } else {
      throw new PizzeriaError({
        name: "RESERVATION_ERROR",
        message: "Table ID is undefined.",
      });
    }
  }
}
