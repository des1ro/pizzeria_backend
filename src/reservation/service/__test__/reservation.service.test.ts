import { PizzeriaError } from "../../../exceptions/pizzeria.exceptions";
import { TableDTO } from "../../table/tableDTO";
import { ReservationService } from "../reservation.service";

describe("Reservation service test suite", () => {
  let objectUnderTest: ReservationService;
  let mockedTableOne: TableDTO;
  let mockedTableTwo: TableDTO;
  let mockedTableThree: TableDTO;
  let mockedTables: Set<TableDTO>;
  beforeEach(() => {
    objectUnderTest = new ReservationService();
  });
  beforeAll(() => {
    mockedTableOne = new TableDTO(1, 2);
    mockedTableTwo = new TableDTO(2, 5);
    mockedTableThree = new TableDTO(3, 7);
  });
  beforeAll(() => {
    mockedTables = new Set<TableDTO>()
      .add(new TableDTO(1, 2))
      .add(new TableDTO(2, 5))
      .add(new TableDTO(3, 7))
      .add(new TableDTO(4, 7))
      .add(new TableDTO(5, 4))
      .add(new TableDTO(6, 6))
      .add(new TableDTO(7, 2))
      .add(new TableDTO(8, 8))
      .add(new TableDTO(9, 4))
      .add(new TableDTO(10, 6));
  });
  describe("Add table to service test suite", () => {
    it("Should add table to service", () => {
      //When
      objectUnderTest.addTable(mockedTableOne);
      objectUnderTest.addTable(mockedTableTwo);
      objectUnderTest.addTable(mockedTableThree);
      //Then
      expect(objectUnderTest).toMatchSnapshot();
    });
    it("Should throw pizzeriaError if table is already in service", () => {
      //When
      objectUnderTest.addTable(mockedTableOne);
      //Then
      expect(() => objectUnderTest.addTable(mockedTableOne)).toThrow(
        PizzeriaError
      );
    });
  });
  it("Should add table to taken and remove from freeTabeles", () => {
    //Given
    const freeTablesSet = new Set<TableDTO>();
    freeTablesSet.add(mockedTableOne).add(mockedTableTwo);
    objectUnderTest = new ReservationService(freeTablesSet);
    //When
    objectUnderTest.bookATable(mockedTableOne);
    //Then
    expect(objectUnderTest).toMatchSnapshot();
  });
  describe("Get a table test suite", () => {
    let freeTablesSet: Set<TableDTO>;
    beforeAll(() => {
      freeTablesSet = new Set<TableDTO>();
      freeTablesSet
        .add(mockedTableOne)
        .add(mockedTableTwo)
        .add(mockedTableThree);
    });
    beforeEach(() => {
      objectUnderTest = new ReservationService(freeTablesSet);
    });
    it("Should find table with right number of seats", () => {
      //When
      const resultOne = objectUnderTest.getATable(2);
      //Then
      expect(resultOne).toBe(mockedTableOne);
    });
    it("Should find most optimal table if there isn't table with exact number of places", () => {
      //When
      objectUnderTest = new ReservationService(mockedTables);
      const resultOne = objectUnderTest.getATable(3);
      //Then
      expect(resultOne.tableId).toBe(5);
    });
    it("Should throw error if there isn't place for so many people", () => {
      //Given
      objectUnderTest = new ReservationService(mockedTables);
      //Then
      expect(() => objectUnderTest.getATable(10)).toThrow(PizzeriaError);
    });
  });
  describe("Set table to available test suite", () => {
    it("Should find table by id and set to available", () => {
      //Given
      objectUnderTest = new ReservationService(
        new Set<TableDTO>(),
        mockedTables
      );
      const mockedTableId = 4;
      //When
      objectUnderTest.setTableToAvailable(mockedTableId);
      //Then
      expect(objectUnderTest).toMatchSnapshot();
    });
    it("Should throw error if tableId is undefined", () => {
      //Then
      expect(() => objectUnderTest.setTableToAvailable(undefined)).toThrow(
        PizzeriaError
      );
    });
    it("Should throw error if table isn't in service", () => {
      //Given
      objectUnderTest = new ReservationService(
        new Set<TableDTO>(),
        mockedTables
      );
      //Then
      expect(() => objectUnderTest.setTableToAvailable(11)).toThrow(
        PizzeriaError
      );
    });
  });
});
