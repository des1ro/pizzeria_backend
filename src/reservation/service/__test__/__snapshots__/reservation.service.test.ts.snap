// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reservation service test suite Add table to service test suite Should add table to service 1`] = `
ReservationService {
  "freeTables": Set {
    TableDTO {
      "seats": 2,
      "tableId": 1,
    },
    TableDTO {
      "seats": 5,
      "tableId": 2,
    },
    TableDTO {
      "seats": 7,
      "tableId": 3,
    },
  },
  "takenTables": Set {},
}
`;

exports[`Reservation service test suite Set table to available test suite Should find table by id and set to available 1`] = `
ReservationService {
  "freeTables": Set {
    TableDTO {
      "seats": 7,
      "tableId": 4,
    },
  },
  "takenTables": Set {
    TableDTO {
      "seats": 2,
      "tableId": 1,
    },
    TableDTO {
      "seats": 5,
      "tableId": 2,
    },
    TableDTO {
      "seats": 7,
      "tableId": 3,
    },
    TableDTO {
      "seats": 4,
      "tableId": 5,
    },
    TableDTO {
      "seats": 6,
      "tableId": 6,
    },
    TableDTO {
      "seats": 2,
      "tableId": 7,
    },
    TableDTO {
      "seats": 8,
      "tableId": 8,
    },
    TableDTO {
      "seats": 4,
      "tableId": 9,
    },
    TableDTO {
      "seats": 6,
      "tableId": 10,
    },
  },
}
`;

exports[`Reservation service test suite Should add table to taken and remove from freeTabeles 1`] = `
ReservationService {
  "freeTables": Set {
    TableDTO {
      "seats": 5,
      "tableId": 2,
    },
  },
  "takenTables": Set {
    TableDTO {
      "seats": 2,
      "tableId": 1,
    },
  },
}
`;
